store/tikv/latch/latch.go

```diff
@@ -33,120 +33,153 @@ type Latch struct {
 }
 
 // NewLatches create a Latches with fixed length,
 // the size will be rounded up to the power of 2.
 func NewLatches(size int) *Latches {
 	powerOfTwoSize := 1 << uint32(bits.Len32(uint32(size-1)))
-	slots := make([]Latch, powerOfTwoSize)
+	slots := make([]latch, powerOfTwoSize)
 	return &Latches{
 		slots:         slots,
-		waitingQueues: make(map[int][]uint64),
+		waitingQueues: make(map[int][]*Lock),
+	}
+}
+
+// genLock generates Lock for the transaction with startTS and keys.
+func (latches *Latches) genLock(startTS uint64, keys [][]byte) *Lock {
+	return &Lock{
+		requiredSlots: latches.genSlotIDs(keys),
+		acquiredCount: 0,
+		startTS:       startTS,
 	}
 }
 
-// GenLock generates Lock for the transaction with startTS and keys.
-func (latches *Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+func (latches *Latches) genSlotIDs(keys [][]byte) []int {
 	slots := make([]int, 0, len(keys))
 	for _, key := range keys {
 		slots = append(slots, latches.slotID(key))
 	}
 	sort.Ints(slots)
 	if len(slots) <= 1 {
-		return NewLock(startTS, slots)
+		return slots
 	}
 	dedup := slots[:1]
 	for i := 1; i < len(slots); i++ {
 		if slots[i] != slots[i-1] {
 			dedup = append(dedup, slots[i])
 		}
 	}
-	return NewLock(startTS, dedup)
+	return dedup
 }
 
 // slotID return slotID for current key.
 func (latches *Latches) slotID(key []byte) int {
 	return int(murmur3.Sum32(key)) & (len(latches.slots) - 1)
 }
 
-// Acquire tries to acquire the lock for a transaction.
-// It returns with stale = true when the transaction is stale(
-// when the lock.startTS is smaller than any key's last commitTS).
-func (latches *Latches) Acquire(lock *Lock) (success, stale bool) {
+// acquire tries to acquire the lock for a transaction.
+func (latches *Latches) acquire(lock *Lock) acquireResult {
+	if lock.IsStale() {
+		return acquireStale
+	}
 	for lock.acquiredCount < len(lock.requiredSlots) {
 		slotID := lock.requiredSlots[lock.acquiredCount]
-		success, stale = latches.acquireSlot(slotID, lock.startTS)
-		if success {
-			lock.acquiredCount++
-			lock.isWaiting = false
-			continue
-		}
-		if !stale {
-			lock.isWaiting = true
+		status := latches.acquireSlot(slotID, lock)
+		if status != acquireSuccess {
+			return status
 		}
-		return
 	}
-	return
+	return acquireSuccess
 }
 
-// Release releases all latches owned by the `lock` and returns the wakeup list.
-// Preconditions: the caller must ensure the transaction is at the front of the latches.
-func (latches *Latches) Release(lock *Lock, commitTS uint64) (wakeupList []uint64) {
-	releaseCount := lock.acquiredCount
-	if lock.isWaiting {
-		releaseCount++
+// release releases all latches owned by the `lock` and returns the wakeup list.
+// Preconditions: the caller must ensure the transaction's status is not locked.
+func (latches *Latches) release(lock *Lock, commitTS uint64) (wakeupList []*Lock) {
+	if lock.isLocked() {
+		panic(fmt.Sprintf("Cann't release locked transaction:%#v", lock))
```

Comment:
I don't think we need the check here. The scheduler returns a `Lock` from the package only after it is not locked:
```
func (scheduler *LatchesScheduler) Lock(startTS uint64, keys [][]byte) *Lock {
	lock := scheduler.latches.genLock(startTS, keys)
	lock.wg.Add(1)
	if scheduler.latches.acquire(lock) == acquireLocked {
		lock.wg.Wait()
	}
	if lock.isLocked() {
		panic("should never run here")
	}
	return lock
}
```

And `release()` can only be called by sending a `Lock` from outside to the scheduler. Therefore the `lock` is guaranteed not locked here.